File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;

class Solution {
    public int[] solution(String[] operations) {
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());

        for (String op : operations) {
            String[] parts = op.split(" ");
            String command = parts[0];
            int value = Integer.parseInt(parts[1]);

            if (command.equals("I")) {
                minHeap.offer(value);
                maxHeap.offer(value);
            } else if (!minHeap.isEmpty()) {
                if (value == 1) {
                    int max = maxHeap.poll();
                    minHeap.remove(max);
                } else {
                    int min = minHeap.poll();
                    maxHeap.remove(min);
                }
            }
        }

        if (minHeap.isEmpty()) {
            return new int[]{0, 0};
        } else {
            return new int[]{maxHeap.peek(), minHeap.peek()};
        }
    }
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq

def solution(operations):
    min_heap = []
    max_heap = []
    number_set = set()

    for operation in operations:
        command, number = operation.split()
        number = int(number)

        if command == 'I':
            heapq.heappush(min_heap, number)
            heapq.heappush(max_heap, -number)
            number_set.add(number)
        elif number_set:
            if number == 1:
                while -max_heap[0] not in number_set:
                    heapq.heappop(max_heap)
                max_value = -heapq.heappop(max_heap)
                number_set.remove(max_value)
            else:
                while min_heap[0] not in number_set:
                    heapq.heappop(min_heap)
                min_value = heapq.heappop(min_heap)
                number_set.remove(min_value)

    if number_set:
        return [max(number_set), min(number_set)]
    else:
        return [0, 0]